src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NuGet.Packaging
+{
+    public class LicenseExpressionParser
+    {
+
+        /// <summary>
+        /// Parses a License Expression if valid.
+        /// The expression would be parsed correct, even if non-standard exceptions are encountered. The non-standard Licenses/Exceptions have metadata on them with which the caller can make decisions.
+        /// Based on the Shunting Yard algorithm. <see href="https://en.wikipedia.org/wiki/Shunting-yard_algorithm"/>
+        /// This method first creates an postfix expression by separating the operators and operands.
+        /// Later the postfix expression is evaluated into an object model that represents the expression. Note that brackets are dropped in this conversion and this is not round-trippable.
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="infixTokens"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the expression has invalid characters.</exception>
+        /// <exception cref="ArgumentException">If the expression itself is invalid. Example: MIT OR OR Apache-2.0, or the MIT or Apache-2.0, because the expressions are case sensitive.</exception>
```

Comment:
Is this list supposed to be exhaustive? If so, `LicenseExpressionTokenizer` throws an `ArgumentNullException` if the string is null or whitespace.

Also, I think the Gallery will want a non-throwing API to verify that the expression is valid. Something like `public static bool TryParse(string expression, out NuGetLicenseExpression licenseExpression)`

Reply:
That's a good point. Maybe instead of throwing you could return a result object that contains the result type (Accepted, MismatchedParentheses, etc..) and the parsed `LicenseExpression`, if any. The Gallery uses this kind of pattern for validation of uploaded nupkgs (see [this type](https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Services/PackageValidationResult.cs) used by [this validation code](https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Services/PackageUploadService.cs#L43)).

Reply:
I don't think that pattern is appropriate for a parser. 
I wouldn't mind that approach if this was a validator, but this is closer to NuGetVersion.Parse than anything else. 
It throws when the license is not a valid expression. 

For example, we will want to warn about certain things(non standard licenses) which are relevant only to the consumer of the APIs, not the parser itself.